### DISCUSSION TITLE
v0.4.0

### DISCUSSION THREAD

**Discussion by Champii** [Champii/Rock#108] [Announcements]
<hr /><em>This discussion was created from the release <a href='https://github.com/Champii/Rock/releases/tag/v0.4.0'>v0.4.0</a>.</em>

**Comment by oraqlle** [comment]
Well done @Champii just tested README against the v0.4.0 and everything works. The new function notation looks really good and it's much easier to tell the function name from its arguments from its rule. 
There does seem to be a small issue with the # comments. The parser also doesn't like empty lines with just tabs and spaces in them to it might be something worth looking into if you're not already. 
Other than that everything is pretty good. Awesome work.

**Reply by Champii** [reply]
Thank you very much ! :)
I also find it more easier to read, I hope that will blend nicely with the closures when they come.
I'm quite happy with the result and the overall direction of the language, I think this might be the last drastic change I made to the syntax. (hopefully lol)
I've also noticed both the bugs you are talking about, the first is because the comments are still half-baked, but there is not much work to do before they stabilize. The second bug is a little more nasty, and is present from a long time. As the language is totally indent-based, the parser uses spaces only to detect indentation. I intend to enforce them and to throw a hard error when a `\t` is found in the source-code.
Also, the parser has to be more relaxed about empty indented lines, as it would always try to parse something after the indentation.
I'll open an issue for both to keep them in mind !

I'm also quite happy that you keep trying the releases as they come out, and I can't thank you enough for that :)
That being said, and if you want to get more involved, I'd be very glad to guide you into the code of the compiler for you to make your first code contribution ! Rock would definitely gain a lot with a second pair of hand and a second set of neurons :p

**Reply by oraqlle** [reply]
It is my pleasure to help, I've been super interested in programming language design for a while know and happy that trying it out has helped you push for continuous progression. I'd would be happy to help out more in the coming week, I've just got University exams to finish and a personal project that is near its final initial release that I've been working on for far too long that I want to finish but I'd be happy to soon. I will continue to test new features as I believe this language to be interesting and unique and want to see where it can go.

**Reply by Champii** [reply]
Oh yeah sure ! I definitely don't want to pressure you, exams are top priority :)
Is that `CortexLib` you are working on ? I've checked it out, it's impressive work.

I think Rock will have less frequent releases, as I have some serious work to do on the trait resolution to allow for generic impl (looking at you, `std::print::Print`). This will take some time and most of the work ahead are even more intense features like `enum` and `closure`.
So don't worry, take your time, and whenever you are ready I'll be your guide through this mess :D

**Reply by oraqlle** [reply]
Thanks. 
Yes it is the CortexLib, although, that particular repository is extremely outdated with the direction I moving at the moment with the entire infrastructure. Most of the work is done on individual repos in a public organisation (not a real one, just labelled that way on GitHub) with smaller more modular libraries that can be developed independently of each other. Other words, its a process at the moment. But thank you for your kind words. 
As for traits, yeah they would be a pain especially IO related operations. My a print trait is valid if the output is a string type as this can always be assumed to be printable. Closures and enums are going to be difficult, and I'm assuming you want a more Rust style enum as opposed to a C++ enum or class enum (std::variant fits this role) so I'm not sure how that would go. I have a think about it.
But thanks, I'll let you know.
